Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1913,71 +1943,155 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
     // Undo assumed inversion about Y axis
     CGAffineTransform textTransform = CGAffineTransformScale(textMatrix, 1.0, -1.0);
     CGAffineTransform deviceTransform = CGContextGetUserSpaceToDeviceSpaceTransform(this);
+    CGAffineTransform combinedTransform = CGAffineTransformConcat(textTransform, deviceTransform);
 
     // Snap the vertical baseline to the nearest pixel to avoid blurring on devices with vertical antialiasing
     textTransform.ty = std::round(textTransform.ty);
 
-    CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, deviceTransform);
-    if ((fabs(finalTextTransform.a * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.b * glyphRun->fontEmSize) <= c_glyphThreshold) ||
-        (fabs(finalTextTransform.d * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.c * glyphRun->fontEmSize) <= c_glyphThreshold)) {
-        TraceWarning(TAG, L"Glyphs too small to be rendered");
-
-        // Not a failure state! Not drawing the glyphs is *okay*.
-        return S_FALSE;
+    bool fastPathTextDrawing =
+        ((textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) ||
+         !transformByGlyph) &&
+        state.textDrawingMode == kCGTextFill;
+
+    std::vector<GlyphRunData> runs;
+
+    size_t maxGlyphCount = std::max_element(glyphRuns, glyphRuns + runsCount, [](const GlyphRunData& left, const GlyphRunData& right) {
+                               return left.run->glyphCount < right.run->glyphCount;
+                           })->run->glyphCount;
+
+    // Shared between transformed glyph runs
+    std::unique_ptr<CGFloat[]> zeroAdvances(new CGFloat[maxGlyphCount]());
```

Comment:
I'm not sure there are major gains by using a unique_ptr rather than vector in this instance, just thought it would be clearer since we don't need to iterate over it or use any other vector functionality